### DISCUSSION TITLE
Verifying Shadowing Behavior

### DISCUSSION THREAD

**Discussion by sno2** [gpuweb/gpuweb#4240] [Q&A] [answered]
I am just trying to verify some behavior regarding shadowing and function resolution. Here are my notes for a few of the shady behaviors.

```wgsl
fn vec2() -> u32 {
	return 0;
}

// This should not affect any of the code.
alias vec2 = u32;

fn main() {
	// This would resolve 'a' to `u32`
	let a = vec2();

	// This would fail - `vec2` resolves to `fn() -> u32`
	// let b = vec2(0, 1);

	// This would resolve 'c' to `vec2<u32>`
	let c = vec2<u32>();

	// This would resolve 'd' to `vec2<u32>`
	var d: vec2<u32>;
}
```

Does this look correct? Thanks for any feedback.

**Comment by jrprice** [comment, marked as answer]
You've declared `vec2` twice at module scope, so that module will be rejected.

If you take out the type alias, your comments for `a` and `b` look correct, but `c` and `d` would fail as `vec2` would resolve to the function, which does not take template arguments.

**Reply by sno2** [reply]
Thank you for clarification. I hadn't realized that it was the same declaration scope for types and everything else.
